kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,750 @@
+
+  if #processed_answers == 0 then
+    if not cname_answer then
+      return {
+        errcode = EMPTY_RECORD_EC,
+        errstr  = EMPTY_RECORD_ESTR,
+        ttl     = self.empty_ttl,
+        -- expire = now() + self.empty_ttl,
+      }
+    end
+
+    table_insert(processed_answers, cname_answer)
+  end
+
+  processed_answers.expire = now() + ttl
+  processed_answers.ttl = ttl
+
+  return processed_answers
+end
+
+
+local function resolve_query(self, name, qtype, tries)
+  local key = name .. ":" .. qtype
+  stats_count(self.stats, key, "query")
+
+  local r, err = resolver:new(self.r_opts)
+  if not r then
+    return nil, "failed to instantiate the resolver: " .. err
+  end
+
+  local options = { additional_section = true, qtype = qtype }
+  local answers, err = r:query(name, options)
+  if r.destroy then
+    r:destroy()
+  end
+
+  if not answers then
+    stats_count(self.stats, key, "query_fail_nameserver")
+    return nil, "DNS server error: " .. (err or "unknown")
+  end
+
+  answers = process_answers(self, name, qtype, answers)
+
+  stats_count(self.stats, key, answers.errstr and "query_fail:" .. answers.errstr
+                                               or "query_succ")
+
+  return answers, nil, answers.ttl
+end
+
+
+local function stale_update_task(premature, self, key, name, qtype, short_key, ttl)
+  if premature then
+    return
+  end
+
+  local answers = resolve_query(self, name, qtype, {})
+  if not answers then
+    -- retry update after failure
+    local retry_delay = math_random(STALE_UPDATE_DELAY, STALE_UPDATE_DELAY * 2)
+    ttl = ttl - retry_delay
+    if ttl < 0 then
+      return  -- no need to retry if it exceeds the stale_ttl
+    end
+    local ok, err = timer_at(retry_delay, stale_update_task, self, key, name,
+                             qtype, short_key, ttl)
+    if not ok then
+      log(ALERT, "failed to start a timer to re-update stale DNS records: ", err)
+    end
+    return
+  end
+
+  if not answers.errcode or answers.errcode == NAME_ERROR_CODE then
+    self.cache:set(key, { ttl = answers.ttl }, answers)
+    insert_last_type(self.cache, name, qtype)
+
+    -- simply invalidate it and let the search iteration choose the correct one
+    self.cache:delete(short_key)
+  end
+end
+
+
+local function start_stale_update_task(self, key, name, qtype, short_key, ttl)
+  stats_count(self.stats, key, "stale")
+
+  local ok, err = timer_at(0, stale_update_task, self, key, name, qtype,
+                           short_key, ttl)
+  if not ok then
+    log(ALERT, "failed to start a timer to update stale DNS records: ", err)
+  end
+end
+
+
+local function resolve_name_type_callback(self, name, qtype, opts, tries)
+  local key = name .. ":" .. qtype
+
+  -- `:peek(stale=true)` verifies if the expired key remains in L2 shm, then
+  -- initiates an asynchronous background updating task to refresh it.
+  local ttl, _, answers = self.cache:peek(key, true)
+  if answers and ttl and not answers.expired then
```

Comment:
qirui: implement lock mechanism  first cache get()

Reply:
```
resolve
  -> get()
   -> answers.expire answer.stale_expire, now() -==> ? stale -
         > start_stale_update_task()
             -> try_lock() -> tiemr-> query
                                    -> not get lock -> ....
```

Reply:
Hi @windmgc Currently, we won't implement a lock mechanism to actively control stale record updates. Instead, we'll continue to let mlcache get the stale data through `peek()->get_stale()`. This implementation is similar to mlcache `resurrect_ttl` and should pose no significant issues. If there are frequent `shdict:set` operations causing expired data to be purged, it also indicates that the DNS server access is active. We can still update expired data through the DNS server in such cases.